ADR/XXXX-pipeline-caching-feature-flag.md

```diff
@@ -0,0 +1,79 @@
+
+* Enable or disable HTTP proxy use for individual pipelines without knowing technical details about the proxy configuration
+* Control proxy usage at the namespace level for all pipelines running in that namespace
+* Control proxy usage at the cluster level for all pipelines in the cluster
+* Apply GitOps processes for this configuration, similar to other Konflux configuration elements
+
+The current design requires users to manually configure `HTTP_PROXY` and `NO_PROXY` parameters to the buildah task, which exposes implementation details and creates a poor user experience. We need a more user-friendly configuration mechanism that abstracts away the technical details while providing appropriate levels of control.
+
+## Decision
+
+We will implement a hierarchical feature flag system for controlling pipeline caching proxy usage with the following configuration levels and precedence:
+
+### Configuration Levels
+
+1. **Pipeline Level**: An `enable-cache-proxy` parameter that can be set in individual pipeline definitions
+2. **Namespace Level**: An `enable-cache-proxy` value in the `konflux-config` config map in the pipeline's namespace
+3. **Cluster Level**: An `enable-cache-proxy` value in the `cluster-config` config map in the `konflux-info` namespace
+
+### Precedence Order
+
+Configuration takes effect based on the following precedence (highest to lowest):
+
+1. Cluster-level configuration (`cluster-config` in `konflux-info` namespace)
+2. Namespace-level configuration (`konflux-config` in pipeline's namespace)
+3. Pipeline-level configuration (`ENABLE_CACHE_PROXY` parameter)
+
+The motivation for this seemingly unintuitive precedence order is to allow for
+overriding the per-pipeline configuration from the namespace or cluster level.
+This serves as a quick "emergency escape hatch" in case of proxy-based failure
+scenarios. See the "Configuration Logic" section below for details.
+
+### Implementation Details
+
+* **Pipeline Integration**: The `ENABLE_CACHE_PROXY` parameter will be passed to the pipeline init task
+* **Configuration Resolution**: The init task will read configuration values from the appropriate config maps and resolve the final proxy settings
+* **Environment Variables**: The init task will emit `HTTP_PROXY` and `NO_PROXY` configuration values to be used by the buildah task
+* **Default Values**: If no configuration is found, the system will fall back emitting empty values, effectively switching the proxy off.
+* **Configuration Values**: Each level supports `true`, `false`, `""` (empty), `"defer"`, or unset values
+* **Proxy Activation**: The proxy will be enabled if any of the configuration levels is set to `true`, following the precedence order above. The emitted values will be:
+  * `HTTP_PROXY`: `squid.caching.svc.cluster.local:3128`
+  * `NO_PROXY`: *&lt;TBD&gt;*
```

Comment:
How NO_PROXY values will be defined? On cluster level? will we allow users to customize them?

You need to export also HTTPS_PROXY, because HTTP_PROXY is used only for HTTP traffic, which should be none.
Also variations of envvars with lowercase must be exported, so all tools are compatible.

Reply:
I've added text to spell-out the distinction between the pipeline and task parameters and variables this ADR concerns itself with and the actual environment variable that are controlled by internal task code.